tests/test_graph_utils.py

```diff
@@ -0,0 +1,71 @@
+import unittest
+from mlpipeline.pipeline import Component, GraphUtils
+
+
+class TestGraphUtils(unittest.TestCase):
+    def test_create_graph(self):
+        components = {
+            "a": Component("a", {"inputs": ["input0", "input1"], "outputs": ["output0"]}),
+            "b": Component("b", {"inputs": ["input0", "input1"], "outputs": ["output1"]}),
+            "c": Component("c", {"inputs": ["a.output0", "b.output1"], "outputs": ["output1"]}),
+        }
+        graph_utils = GraphUtils()
+        expected_graph = {"inputs": {"b", "a"}, "b": {"c"}, "a": {"c"}, "c": {"outputs"}, "outputs": set()}
+        output_graph = graph_utils._create_graph(components, set(["input0", "input1"]), set(["c.output1"]))
+        self.assertEqual(expected_graph, output_graph)
+
+        components = {
+            "a": Component("a", {"inputs": [], "outputs": ["output0"]}),
+            "c": Component("c", {"inputs": ["a.output0"], "outputs": ["output1"]}),
+        }
+        expected_graph = {"a": {"c"}, "c": {"outputs"}, "outputs": set()}
+        output_graph = graph_utils._create_graph(components, set(), set(["c.output1"]))
+        self.assertEqual(expected_graph, output_graph)
+
+    def test_get_topological_order(self):
+        graph = {
+            "inputs": set(["a", "b"]),
+            "a": set(["c"]),
+            "b": set(["c"]),
+            "c": set(["outputs"]),
+            "outputs": set([])
+        }
+        graph_utils = GraphUtils()
+        running_order = graph_utils._get_topological_order(graph)
+        expected_running_order_reduced = [None, None, "c"]
+        self.assertEqual(expected_running_order_reduced[2], running_order[2])
+
+        graph = {
+            "inputs": set(["a", "b"]),
+            "a": set(["b"]),
+            "b": set(["c"]),
+            "c": set(["a"]),
+            "outputs": set([])
+        }
+        with self.assertRaises(RuntimeError):
+            graph_utils._get_topological_order(graph)
```

Comment:
I've been bitten badly by assertRaises on such general exceptions. Asserting at least part of the exception string would make the test stronger.

Reply:
Totally agree. I could also implement my own exceptions, but didn't think of it at all. Thanx